executor/src/witgen/expression_evaluator.rs

```diff
@@ -23,71 +23,96 @@ pub trait SymbolicVariables<T> {
     }
 }
 
-pub struct ExpressionEvaluator<T, SV> {
-    variables: SV,
-    marker: PhantomData<T>,
+pub struct ExpressionEvaluator<'a, T> {
+    intermediate_definitions: &'a BTreeMap<PolyID, &'a Expression<T>>,
+    /// Maps intermediate polynomial IDs to their evaluation. Updated throughout the lifetime of the
+    /// ExpressionEvaluator.
+    intermediates_cache: BTreeMap<PolyID, AffineResult<AlgebraicVariable<'a>, T>>,
 }
 
-impl<T, SV> ExpressionEvaluator<T, SV>
-where
-    SV: SymbolicVariables<T>,
-    T: FieldElement,
-{
-    pub fn new(variables: SV) -> Self {
+impl<'a, T: FieldElement> ExpressionEvaluator<'a, T> {
+    pub fn new(intermediate_definitions: &'a BTreeMap<PolyID, &'a Expression<T>>) -> Self {
         Self {
-            variables,
-            marker: PhantomData,
+            intermediate_definitions,
+            intermediates_cache: Default::default(),
         }
     }
+
     /// Tries to evaluate the expression to an affine expression in the witness polynomials
     /// or publics, taking their current values into account.
+    /// Might update its cache of evaluations of intermediate polynomials.
     /// @returns an expression affine in the witness polynomials or publics.
-    pub fn evaluate<'a>(&self, expr: &'a Expression<T>) -> AffineResult<AlgebraicVariable<'a>, T> {
+    pub fn evaluate<SV: SymbolicVariables<T>>(
+        &mut self,
+        variables: &SV,
```

Comment:
It's a bit weird that we don't receive the variables in the constructor anymore.

The reason is that `RowPair` wants to use the same evaluator instance to benefit more from the intermediates cache, so it needs to own it. But `RowPair` also implements `WitnessColumnEvaluator`, which is used by `SymbolicWitnessEvaluator`. So it would have to own an evaluator that holds a reference to itself, which is not possible AFAIK.